// continuityfactory/continuityfactory-util-space-export.js
/*
	Factory for export space data template
*/

var mydigitalstructure = require('mydigitalstructure')
var _ = require('lodash')
var moment = require('moment');
const { remove } = require('lodash');

module.exports = 
{
	VERSION: '0.0.1',

	init: function (param)
	{
		//https://attacomsian.com/blog/nodejs-write-json-object-to-file

		mydigitalstructure.add(
		{
			name: 'continuity-util-space-export-data',
			code: function (param)
			{
				var event = mydigitalstructure.get({scope: '_event'});

				console.log(event.exports);

				if (event.exports == undefined)
				{
					console.log('Event has no exports defined (event.exports).')
				}
				else
				{
					mydigitalstructure.set(
					{
						scope: 'continuity-util-space-export-data-process',
						context: 'index',
						value: 0
					});

					mydigitalstructure.set(
					{
						scope: 'continuity-util-space-export-data-process',
						context: 'template-file-data',
						value: {}
					});

					mydigitalstructure.invoke('continuity-util-space-export-data-process');
				}
			}
		});

		mydigitalstructure.add(
		{
			name: 'continuity-util-space-export-data-process',
			code: function (param)
			{
				var event = mydigitalstructure.get({scope: '_event'});

				var index = mydigitalstructure.get(
				{
					scope: 'continuity-util-space-export-data-process',
					context: 'index'
				});

				if (index < event.exports.length)
				{
					var exportSearch = event.exports[index];
					exportSearch = _.assign(exportSearch,
					{
						callback: 'continuity-util-space-export-data-process-next',
						callbackParam: param
					});

					if (exportSearch.rows == undefined) {exportSearch.rows = 999}

					mydigitalstructure.cloud.search(exportSearch);
				}
				else
				{
					if (event.template == 'true')
					{
						var templateData = mydigitalstructure.get(
						{
							scope: 'continuity-util-space-export-data-process',
							context: 'template-file-data'
						});

						fileDataTemplate =
						{
							template: templateData
						}

						filename = 'data/setup-template-' + moment().format('DD-MMM-YYYY-HH-mm') + '.json'

						mydigitalstructure.invoke('util-save-to-file',
						{
							filename: filename,
							fileData: fileDataTemplate,
							onComplete: 'continuity-util-space-export-data-process-complete'
						});
	
					}
					else
					{
						mydigitalstructure.invoke('continuity-util-space-export-data-process-complete')
					}
				}
			}
		});

		mydigitalstructure.add(
		{
			name: 'continuity-util-space-export-data-process-next',
			code: function (param, response)
			{
				var event = mydigitalstructure.get({scope: '_event'});
				var index = mydigitalstructure.get(
				{
					scope: 'continuity-util-space-export-data-process',
					context: 'index'
				});

				mydigitalstructure.set(
				{
					scope: 'continuity-util-space-export-data-process',
					context: 'index',
					value: index + 1
				});

				var exportSearch = event.exports[index];

				if (response.status == 'ER')
				{
					console.log('!!ERROR; Bad Search');
					console.log(exportSearch);
				}
				else
				{
					var fileData = response.data.rows;
					if (event.removeID == 'true')
					{
						_.each(fileData, function (data)
						{
							delete data.id
						});
					}

					if (event.template == 'true')
					{
						var templateData = mydigitalstructure.get(
						{
							scope: 'continuity-util-space-export-data-process',
							context: 'template-file-data'
						});

						templateData[exportSearch.object] = fileData;

						mydigitalstructure.invoke('continuity-util-space-export-data-process')
					}
					else
					{
						var filename = 'data/' + _.kebabCase(exportSearch.object) + '.json';
		
						mydigitalstructure.invoke('util-save-to-file',
						{
							filename: filename,
							fileData: fileData,
							onComplete: 'continuity-util-space-export-data-process'
						});
					}
				}
			}
		});

		mydigitalstructure.add(
		{
			name: 'continuity-util-space-export-data-process-complete',
			code: function (param, response)
			{
				var event = mydigitalstructure.get({scope: '_event'});
				mydigitalstructure.invoke('util-end', 'Export complete. [' + event.exports.length + ']');
			}
		});
	}
}
